src/TypeAdapter/ArrayTypeAdapter.php

```diff
@@ -0,0 +1,210 @@
+<?php
+/*
+ * Copyright (c) Nate Brunette.
+ * Distributed under the MIT License (http://opensource.org/licenses/MIT)
+ */
+
+declare(strict_types=1);
+
+namespace Tebru\Gson\TypeAdapter;
+
+use LogicException;
+use Tebru\Gson\Exception\JsonSyntaxException;
+use Tebru\Gson\Internal\TypeAdapterProvider;
+use Tebru\Gson\Context\ReaderContext;
+use Tebru\Gson\TypeAdapter;
+use Tebru\Gson\Context\WriterContext;
+use Tebru\PhpType\TypeToken;
+
+/**
+ * Class ArrayTypeAdapter
+ *
+ * @author Nate Brunette <n@tebru.net>
+ */
+class ArrayTypeAdapter extends TypeAdapter
+{
+    /**
+     * @var TypeAdapterProvider
+     */
+    protected $typeAdapterProvider;
+
+    /**
+     * @var TypeToken
+     */
+    protected $keyType;
+
+    /**
+     * @var TypeAdapter
+     */
+    protected $valueTypeAdapter;
+
+    /**
+     * @var int
+     */
+    protected $numberOfGenerics;
+
+    /**
+     * A TypeAdapter cache keyed by raw type
+     *
+     * @var TypeAdapter[]
+     */
+    protected $adapters = [];
+
+    /**
+     * Constructor
+     *
+     * @param TypeAdapterProvider $typeAdapterProvider
+     * @param TypeToken $keyType
+     * @param TypeAdapter $valueTypeAdapter
+     * @param int $numberOfGenerics
+     */
+    public function __construct(
+        TypeAdapterProvider $typeAdapterProvider,
+        TypeToken $keyType,
+        TypeAdapter $valueTypeAdapter,
+        int $numberOfGenerics
+    ) {
+        $this->typeAdapterProvider = $typeAdapterProvider;
+        $this->keyType = $keyType;
+        $this->valueTypeAdapter = $valueTypeAdapter;
+        $this->numberOfGenerics = $numberOfGenerics;
+    }
+
+    /**
+     * Read the next value, convert it to its type and return it
+     *
+     * @param array|null $value
+     * @param ReaderContext $context
+     * @return array|null
+     */
+    public function read($value, ReaderContext $context): ?array
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        if (!is_array($value)) {
+            throw new JsonSyntaxException(sprintf('Could not parse json, expected array or object but found "%s"', gettype($value)));
+        }
+
+        $result = [];
+
+        if ($this->numberOfGenerics > 2) {
+            throw new LogicException('Array may not have more than 2 generic types');
+        }
+
+        if ($this->keyType->phpType !== TypeToken::WILDCARD
+            && $this->keyType->phpType !== TypeToken::STRING
+            && $this->keyType->phpType !== TypeToken::INTEGER
+        ) {
+            throw new LogicException('Array keys must be strings or integers');
+        }
+
+        $arrayIsObject = $this->numberOfGenerics === 2 || is_string(key($value));
+        $enableScalarAdapters = $context->enableScalarAdapters();
+
+        foreach ($value as $key => $item) {
+            $itemValue = null;
+            switch ($this->numberOfGenerics) {
+                case 0:
+                    if (!$enableScalarAdapters && is_scalar($item)) {
```

Comment:
So if scalar adapters are not being used, then just grab the value directly from the item rather than reading it through the adapter right?

Reply:
Yes, it's unfortunate there's not an easy way to cast it to a variable type. It's a risk I'm willing to take for now. If wrong types become a problem, I'll revisit.